vowpalwabbit/cb.cc

```diff
@@ -281,11 +285,11 @@ void parse_label(parser* p, shared_data* sd, void* v, v_array<VW::string_view>&
 
   ld->action = (uint32_t)hashstring(words[0].begin(), words[0].length(), 0);
 
-  words.begin()++;
-
+  // Removing the first element of a vector is not efficient at all, every element must be copied/moved.
```

Comment:
I am not familiar with the `words` vector/array, but from a quick search it doesn't seem to have index lookups, or if it does they don't go past the 3rd index. Would it be worth making it a linked list instead of a vector?

Reply:
The words vector (like you said) is often accessed randomly, even if it is only to around the 3rd index. It looks to me like `std::list` doesn't allow for random access?

Reply:
Discussed offline and std::list doesn't make sense because the container is always small in length.

Reply:
Doesn't the previous code remove the first one, then call parse_label? or am i missing something

Reply:
```
[xxxx]
 ^begin
 
>> words.begin()++;

[xxxx]
  ^begin

>> CB::parse_label(p, sd, &(ld->event), words);
>> words.begin()--;

[xxxx]
 ^begin
```

It doesn't remove it, it moves the begin pointer of the buffer so that to `parse_label` it doesn't exist. This is of course very unsafe and not something other containers really let you do. It was able to happen because currently (to be fixed) `v_array` returns a reference to it's own begin pointer from `begin`